### PR TITLE
char: silabs: si446x (SiLabs Si446x Transceiver) driver added. 

### PR DESCRIPTION
**Pull request by sunipkm**
Implements basic API calls over ioctl, and receive/transmit using read/write. Reads can be non-blocking with polling support. Writes are blocking. Overlay `si446x-spi0` provided.
The device is initialized by an userspace program that requires inclusion of the radio configuration generated by the WDS tool from SiLabs, and patched following Zak Kemble's guide (https://blog.zakkemble.net/si4463-radio-library-avr-arduino/).

### CODE REVIEW

**Comment by pelwell** [comment]
I'll accept the overlay (with an appropriate README entry), but the driver needs to be submitted upstream to the Linux kernel maintainers.

**Comment by sunipkm** [comment]
I have removed the driver, and updated the README for the overlay as requested.

File: arch/arm/boot/dts/overlays/README [hunk 1]
```diff
@@ -2720,6 +2720,12 @@ Params: speed                   SPI bus speed (default 4000000)
         reset_pin               GPIO pin for RESET (default 25)
         height                  Display height (32 or 64; default 64)
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README:2722]: Two blank lines between overlays, please.

File: arch/arm/boot/dts/overlays/README [hunk 2]
```diff
@@ -2720,6 +2720,12 @@ Params: speed                   SPI bus speed (default 4000000)
...
 
+Name:   si446x-spi0
+Info:   Overlay for Si446x UHF Transceiver via SPI using si446x driver (currently out-of-tree at https://github.com/sunipkmukherjee/silabs.git)
+Load:   dtoverlay=si446x-spi0,<param>=<val>
+Params: speed                   SPI bus speed (default 4000000)
+        int_pin                 GPIO pin for interrupts (default 17)
+        reset_pin               GPIO pin for RESET (default 27)
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README:2731]: Another blank line here.

File: arch/arm/boot/dts/overlays/README [hunk 3]
```diff
@@ -2720,6 +2720,12 @@ Params: speed                   SPI bus speed (default 4000000)
         reset_pin               GPIO pin for RESET (default 25)
         height                  Display height (32 or 64; default 64)
 
+Name:   si446x-spi0
+Info:   Overlay for Si446x UHF Transceiver via SPI using si446x driver (currently out-of-tree at https://github.com/sunipkmukherjee/silabs.git)
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/README]: This line is far too long - there is a hard limit of 80 characters in this file.

File: arch/arm/boot/dts/overlays/si446x-spi0-overlay.dts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+// Overlay for the SiLabs Si446X Controller - SPI0
+// Default Interrupt Pin: 17
+// Default SDN Pin: 27
+/dts-v1/;
+/plugin/;
+
+   / {
+    compatible = "brcm,bcm2708";
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/si446x-spi0-overlay.dts]: This should be `"brcm,bcm2835"`.

File: arch/arm/boot/dts/overlays/si446x-spi0-overlay.dts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            };
+        };
+    };
+
+    __overrides__ {
+        int_pin = <&uhf0>, "interrupts:0",
+                  <&uhf0>, "irq_pin:0",
+                  <&uhf0_pins>, "brcm,pins:0";
+        reset_pin = <&uhf0>, "sdn_pin:0",
+                    <&uhf0_pins>, "brcm,pins:1";
```
Comment (pelwell) [inline arch/arm/boot/dts/overlays/si446x-spi0-overlay.dts]: The offset for a parameter target is in bytes - this should say `"brcm,pins:4"`.

**Review by pelwell — COMMENTED** [review]
There are a few small issues with the overlay and README entry that I've commented on, but the real problem with this PR is that without the driver people can't use it, and if they're building their own kernel/module they might as well build the overlay with it.

**Comment by sunipkm** [comment]
> There are a few small issues with the overlay and README entry that I've commented on, but the real problem with this PR is that without the driver people can't use it, and if they're building their own kernel/module they might as well build the overlay with it.

I agree with you that anyone using the driver can build their own overlay as well. I have submitted a message to LKML, and waiting for a follow up. In the mean time, the inclusion of the overlay can point people towards the code, and independent testing that will be very helpful. The available info/code about this transceiver is not large, which is why I am looking into getting this to a place that can benefit others.

Thanks for the corrections on the overlay.

**Comment by pelwell** [comment]
Fixed up and merged offline.

**Comment by pelwell** [comment]
Did you get anywhere with the upstreaming? Some checking has just flagged `silabs,si446x` as a compatible string that isn't implemented by any drivers in our trees.

**Comment by sunipkm** [comment]
Hi Phil,

Sorry to say it's been in development hell for a bit while I've been
working on my PhD, I'll be going back to it in a next couple of months.
There is a memory buffer bug in the implementation currently that I need to
fix.

Hope you are doing well.

Sunip

**Comment by pelwell** [comment]
No problem - I'll add the compatible string to the list of exceptions.

**Pull request closed without merging**